wayland-client/src/async_event_queue.rs

```diff
@@ -0,0 +1,183 @@
+use std::{
+    io,
+    num::NonZeroUsize,
+    sync::{
+        Arc,
+        atomic::Ordering,
+    },
+    os::fd::{
+        AsFd,
+        BorrowedFd,
+    },
+};
+use crate::{
+    conn::SyncData,
+    Connection,
+    EventQueue,
+    DispatchError,
+    WaylandError,
+    protocol::{
+        wl_display,
+    },
+};
+use tokio::io::{
+    unix::AsyncFd,
+    Ready,
+};
+
+/// Asynchronous version of [`EventQueue`], taking borrowed references to an [`EventQueue`],
+/// allowing for non-blocking polled asynchronous operations against the Queue using
+/// [`AsyncFd`] from the tokio runtime to poll against the wayland socket
+#[derive(Debug)]
+pub struct AsyncEventQueue<'a, State> {
+    /// Inner reference to synchronous version of this event queue
+    pub queue: EventQueue<State>,
+    conn: &'a Connection,
+    afd: AsyncFd<BorrowedFd<'a>>,
+}
+
+impl<'a, State> AsyncEventQueue<'a, State> {
+    /// Creates a new [`AsyncEventQueue`] from a reference to a given [`Connection`] and
+    /// *synchronous* [`EventQueue`]
+    #[inline]
+    pub fn new(
+        conn: &'a Connection,
+        event_queue: EventQueue<State>,
+    ) -> io::Result<Self> {
+        Ok(AsyncEventQueue {
+            queue: event_queue,
+            conn: conn,
+            afd: AsyncFd::new(conn.as_fd())?,
+        })
+    }
+
+    #[inline(always)]
+    fn dispatch_pending(&mut self, data: &mut State) -> Result<Option<NonZeroUsize>, DispatchError> {
+        self.queue.dispatch_pending(data)
+            .map(NonZeroUsize::new)
+    }
+
+    /// Asynchronous version of [`EventQueue::blocking_dispatch`] that uses tokio's fd-based
+    /// polling system to avoid blocking.
+    ///
+    /// This function completes once any number of events have been dispatched *on this thread*.
+    ///
+    /// In the success case, returns the # of events that were dispatched.
+    pub async fn dispatch_single(&mut self, data: &mut State) -> Result<usize, DispatchError> {
+        self.queue.flush_d()?;
+        if let Some(v) = self.dispatch_pending(data)? {
+            return Ok(v.get());
+        }
+        loop {
+            let mut guard = self.afd.readable_mut().await
+                .map_err(|e| DispatchError::Backend(WaylandError::from(e)))?;
+            let lock = if let Some(v) = self.queue.prepare_read() {
+                v
+            } else {
+                if let Some(v) = self.dispatch_pending(data)? {
+                    return Ok(v.get());
+                } else {
+                    continue;
+                }
+            };
+            let ret = match lock.read() {
+                Ok(..) => Some(self.queue.dispatch_pending(data)),
+                Err(ref e) if e.would_block() => {
+                    guard.clear_ready_matching(Ready::READABLE);
+                    None
+                },
+                Err(e) => Some(Err(DispatchError::from(e))),
+            };
```

Comment:
This implementation is not correct, and it may be difficult to be made correct given the tokio paradigm.

The way `prepare_read()` work (and this is a constraint coming from libwayland itself), it must be invoked *before* waiting for the FD to become readable. Not doing so will cause issues in a multithreaded context, which some threads never being made aware that events are available to process.

At the same time, because the guard returned by `prepare_read()` behaves like a semaphore, you also need to ensure that never more than one such guard is created on a single thread (otherwise this will cause a deadlock), and that the guard is as short-lived as possible, ie it should be be created just before starting to wait on epoll, and consume just after epoll has returned (by invoking `.read()` or by dropping it). Having a thread hold onto such a guard while doing other stuff will block all other threads from reading from the socket, and may in turn cause deadlocks on the app at large.

I'm bringing these points not as a "worst case scenario" hypothetical, those are issues that we have already encountered previously.

Reply:
> The way `prepare_read()` work (and this is a constraint coming from libwayland itself), it must be invoked _before_ waiting for the FD to become readable. Not doing so will cause issues in a multithreaded context, which some threads never being made aware that events are available to process.

Unfortunate. I had done some *super* basic testing and it had seemed okay, but that makes sense that that requirement would completely negate this implementation.

Thanks for the info; definitely the reason I listed this implementation as one of the things that needed looked over from someone more intimately familiar than me.